### PR TITLE
refactor accelerator normalization

### PR DESCRIPTION
**Pull request by guotuofeng**
## Describe your changes
* Refactor accelerator normalization logic so that the host system could reuse it when we introduce host device later.
* If user specify EP and device explicitly, the EP will always be override by the available EPs. This PR also fix the bug. 

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: olive/hardware/accelerator.py [hunk 1]
```diff
@@ -146,12 +146,36 @@ def infer_accelerators_from_execution_providers(execution_providers: List[str]):
...
 
-def infer_accelerators_for_system_config(system_config: "SystemConfig") -> "SystemConfig":
-    """If the accelerators are not specified, infer the accelerators. otherwise, return the original system config."""
+        if execution_providers == ["CPUExecutionProvider"]:
+            inferred_accelerators = ["cpu"]
+        else:
+            inferred_accelerators = AcceleratorLookup.infer_devices_from_execution_providers(execution_providers)
+            assert (
+                inferred_accelerators
+            ), f"Cannot infer the accelerators from the execution providers {execution_providers}."
```
Comment (trajepl) [inline olive/hardware/accelerator.py:160]: infer the ~~accelerators~~ devices? Align with the config name?

File: olive/hardware/accelerator.py [hunk 2]
```diff
@@ -146,12 +146,36 @@ def infer_accelerators_from_execution_providers(execution_providers: List[str]):
...
+            assert len(inferred_accelerators) == 1, (
+                f"Cannot infer the accelerators from the execution providers {execution_providers}. "
```
Comment (trajepl) [inline olive/hardware/accelerator.py:162]: Ditto. Also please remind user that they can set the device manually to mitigate this issue.

File: olive/hardware/accelerator.py [hunk 3]
```diff
@@ -241,6 +239,9 @@ def infer_accelerators_for_system_config(system_config: "SystemConfig") -> "Syst
                             accelerator.device,
                             accelerator.execution_providers,
                         )
+                    else:
+                        # User specify both the device and execution providers
```
Comment (trajepl) [inline olive/hardware/accelerator.py:243]: This comment can be removed. The debug log can tell same thing.

File: test/unit_test/hardware/test_accelerator.py [hunk 4]
```diff
@@ -175,6 +142,180 @@ def test_create_accelerators(get_available_providers_mock, system_config, expect
...
+                        "CUDAExecutionProvider",
+                        "CPUExecutionProvider",
+                    ],
+                }
+            ],
+            [],
+            ["CUDAExecutionProvider", "CPUExecutionProvider"],
+        ),
+        (
+            # dosn't fill both device and ep.
```
Comment (trajepl) [inline test/unit_test/hardware/test_accelerator.py:220]: -> doesn't

File: olive/hardware/accelerator.py (hunk 1 above)
Comment (guotuofeng) [reply olive/hardware/accelerator.py:160]: sure, will do

**Review by trajepl — APPROVED** [review]

**Pull request merged**